### PR TITLE
GGUF: Load and save metadata

### PR DESCRIPTION
**Pull request by jbochi**
## Proposed changes

Another incomplete follow up for https://github.com/ml-explore/mlx/pull/350. 

Probably higher priority than https://github.com/ml-explore/mlx/pull/426.

The goal is to be able to load and save metadata. For now, it only supports strings.

Primitive types such as the different flavors of ints and floats should be casted to mx.array scalars. 

I've tested that the file generated by the c++ test can also be read by gguf-tools:

```
~/dev/github/gguf-tools/gguf-tools show /var/folders/sw/_m_mwjxn1mlc4vzkj4b150mm0000gn/T/test_arr.gguf
/var/folders/sw/_m_mwjxn1mlc4vzkj4b150mm0000gn/T/test_arr.gguf (ver 3): 1 key-value pairs, 1 tensors
test_str: [string] my string
f32 tensor test @128, 5 weights, dims [5], 20 bytes
```

cc @awni 

## Checklist

Put an `x` in the boxes that apply.

- [ ] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [ ] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/io.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    std::shared_ptr<io::Writer> in_stream,
+    std::unordered_map<std::string, array>);
+void save_safetensors(
+    const std::string& file,
+    std::unordered_map<std::string, array>);
+
+using MetaData =
+    std::variant<std::monostate, array, std::string, std::vector<std::string>>;
+
+/** Load metadata and array map from .gguf file format */
```
Comment (jbochi) [inline mlx/io.h:44]:
```suggestion
/** Load array map and metadata from .gguf file format */
```

Nit: Change order in comment to match what is returned.

File: mlx/io/gguf.cpp [hunk 2]
```diff
@@ -12,6 +15,9 @@ extern "C" {
 
 namespace mlx::core {
 
+// https://github.com/antirez/gguf-tools/blob/main/gguflib.h#L102-L108
```
Comment (jbochi) [inline mlx/io/gguf.cpp:18]:
```suggestion
// https://github.com/antirez/gguf-tools/blob/af7d88d808a7608a33723fba067036202910acb3/gguflib.h#L102-L108
```

Permalink

File: mlx/io/gguf.cpp [hunk 3]
```diff
@@ -70,15 +76,129 @@ std::tuple<allocator::Buffer, Dtype> extract_tensor_data(gguf_tensor* tensor) {
...
+      break;
+    case GGUF_VALUE_TYPE_BOOL:
+      value = array(val->boolval, bool_);
+      break;
+    case GGUF_VALUE_TYPE_STRING:
+      value =
+          std::string(val->string.string, static_cast<int>(val->string.len));
+      break;
+    case GGUF_VALUE_TYPE_FLOAT64:
+      value = array(val->float64, float32);
```
Comment (jbochi) [inline mlx/io/gguf.cpp:120]: Don't we need to cast it to float32?

File: mlx/io/gguf.cpp [hunk 4]
```diff
@@ -12,6 +15,9 @@ extern "C" {
 
 namespace mlx::core {
 
+// https://github.com/antirez/gguf-tools/blob/main/gguflib.h#L102-L108
+constexpr int gguf_array_header = 12;
```
Comment (jbochi) [inline mlx/io/gguf.cpp:19]: When I first read this I thought it was the constant that represents the array type. Perhaps rename it to `gguf_array_header_size` to make it more explicit?

File: mlx/io/gguf.cpp [hunk 5]
```diff
@@ -70,15 +76,129 @@ std::tuple<allocator::Buffer, Dtype> extract_tensor_data(gguf_tensor* tensor) {
...
+            str = std::string(str_val->string, static_cast<int>(str_val->len));
+            ctx->off += (str_val->len + sizeof(gguf_string));
+          }
+          value = std::move(strs);
+          break;
+        }
+        case GGUF_VALUE_TYPE_FLOAT64:
+          value = array(reinterpret_cast<double*>(data), {size}, float32);
+          break;
+      }
```
Comment (jbochi) [inline mlx/io/gguf.cpp:175]:
```suggestion
          default:
            throw std::runtime_error("[load_gguf] Array of unsupported type.");
            break;
      }
```

Raise an error if type is not supported (such as nested arrays).

File: mlx/io/gguf.cpp [hunk 6]
```diff
@@ -70,15 +76,129 @@ std::tuple<allocator::Buffer, Dtype> extract_tensor_data(gguf_tensor* tensor) {
...
+      break;
+    }
+    default:
+      throw std::runtime_error("[load_gguf] Received unexpected type.");
+      break;
+  }
+  if (type == GGUF_VALUE_TYPE_STRING) {
+    ctx->off += (8 + std::get<std::string>(value).size());
```
Comment (jbochi) [inline mlx/io/gguf.cpp:183]: perhaps define a `gguf_string_header_size` for `8`?

File: mlx/io/gguf.cpp [hunk 7]
```diff
@@ -89,27 +209,178 @@ std::unordered_map<std::string, array> load_gguf(
...
+    gguf_val->array.len = val.size();
+    memcpy(val_vec.data() + gguf_array_header, val.data<char>(), val.nbytes());
+    gguf_append_kv(
+        ctx,
+        key.c_str(),
+        key.length(),
+        GGUF_VALUE_TYPE_ARRAY,
+        reinterpret_cast<void*>(val_vec.data()),
+        gguf_size);
+  } else {
```
Comment (jbochi) [inline mlx/io/gguf.cpp:250]: is this ever the case? Reading the code, it seems that we call `gguf_append_kv` directly if we have a string or an array of strings.

File: python/src/load.cpp [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
+#include <iostream> // TODO
```
Comment (jbochi) [inline python/src/load.cpp:5]:
```suggestion
#include <iostream> // TODO
```

I think we can remove this now

File: python/tests/test_load.py [hunk 9]
```diff
@@ -117,6 +117,99 @@ def test_save_and_load_gguf(self):
...
+        self.assertTrue("test" in load_dict)
+        self.assertTrue(mx.array_equal(load_dict["test"], save_dict["test"]))
+
+        # Loads empty metadata
+        load_dict, meta_load_dict = mx.load(save_file_mlx, return_metadata=True)
+        self.assertTrue("test" in load_dict)
+        self.assertTrue(mx.array_equal(load_dict["test"], save_dict["test"]))
+        self.assertEqual(len(meta_load_dict), 0)
+
+        metadata = {"meta": "data"}
```
Comment (jbochi) [inline python/tests/test_load.py:142]:
```suggestion
        # Loads string metadata
        metadata = {"meta": "data"}
```

File: python/tests/test_load.py [hunk 10]
```diff
@@ -117,6 +117,99 @@ def test_save_and_load_gguf(self):
...
+        mx.save_gguf(save_file_mlx, save_dict, metadata)
+        load_dict, meta_load_dict = mx.load(save_file_mlx, return_metadata=True)
+        self.assertTrue("test" in load_dict)
+        self.assertTrue(mx.array_equal(load_dict["test"], save_dict["test"]))
+        self.assertEqual(len(meta_load_dict), 1)
+        self.assertTrue("meta" in meta_load_dict)
+        self.assertEqual(meta_load_dict["meta"], "data")
+
+        for t in [
```
Comment (jbochi) [inline python/tests/test_load.py:151]:
```suggestion
        # Test scalars and one dimensional arrays
        for t in [
```

File: python/tests/test_load.py [hunk 11]
```diff
@@ -117,6 +117,99 @@ def test_save_and_load_gguf(self):
                             mx.array_equal(load_dict["test"], save_dict["test"])
                         )
 
+    def test_save_and_load_gguf_with_metadata(self):
```
Comment (jbochi) [inline python/tests/test_load.py:120]: Nit: Perhaps it's worth breaking this into multiple tests

File: tests/load_tests.cpp [hunk 12]
```diff
@@ -37,33 +37,160 @@ TEST_CASE("test save_safetensors") {
...
 
-  save_gguf(file_path, map);
-  auto loaded = load_gguf(file_path);
-  CHECK_EQ(loaded.size(), 2);
-  CHECK_EQ(loaded.count("test"), 1);
-  CHECK_EQ(loaded.count("test2"), 1);
-  for (auto [k, v] : loaded) {
-    CHECK(array_equal(v, map.at(k)).item<bool>());
+  {
+    // Check saving loading just arrays, no metadata
```
Comment (jbochi) [inline tests/load_tests.cpp:45]:
```suggestion
    // Check saving and loading just arrays, no metadata
```

File: tests/load_tests.cpp [hunk 13]
```diff
@@ -37,33 +37,160 @@ TEST_CASE("test save_safetensors") {
...
+    CHECK_EQ(loaded_metadata.size(), 0);
+    CHECK_EQ(loaded_weights.size(), 2);
+    CHECK_EQ(loaded_weights.count("test"), 1);
+    CHECK_EQ(loaded_weights.count("test2"), 1);
+    for (auto [k, v] : loaded_weights) {
+      CHECK(array_equal(v, original_weights.at(k)).item<bool>());
+    }
+  }
+
+  std::unordered_map<std::string, MetaData> original_metadata;
```
Comment (jbochi) [inline tests/load_tests.cpp:57]:
```suggestion
  # Test saving and loading string metadata
  std::unordered_map<std::string, MetaData> original_metadata;
```

Comment (jbochi) [reply tests/load_tests.cpp:57]: We could also add a test for array of strings

**Review by jbochi — COMMENTED** [review]
Looks great! Thanks for the clean up and finishing this.

If my understanding is correct, 1D GGUF arrays with a single element can be distinguished from raw values because the former is serialized as a 1D mx array and the latter as a scalar MX array, correct? Perhaps we include this in the documentation.

File: mlx/io/gguf.cpp (hunk 3 above)
Comment (jbochi) [reply mlx/io/gguf.cpp:120]: (It's not covered by tests because there's no equivalent mx.float64)

**Comment by jbochi** [comment]
PS: I left a bunch of comments, but most of them are just comments about comments. Code looks great!

Thanks again for finishing the work and putting it in a better shape than I thought would be possible :)

File: mlx/io/gguf.cpp (hunk 3 above)
Comment (awni) [reply mlx/io/gguf.cpp:120]: No std::copy will handle casting this.

File: mlx/io/gguf.cpp (hunk 7 above)
Comment (awni) [reply mlx/io/gguf.cpp:250]: This case is to handle scalars (0 dim) arrays differently than 1 d arrays.

File: mlx/io/gguf.cpp (hunk 6 above)
Comment (awni) [reply mlx/io/gguf.cpp:183]: Good point, used `sizeof(gguf_string)`

**Comment by awni** [comment]
Brilliant, thanks for the comments, those are very nice! All addressed.

I just pushed. When the tests clear, we can merge!! Thanks for making this happen 🚀

**Review by awni — APPROVED** [review]

**Pull request merged**